devito/dle/blocking_utils.py

```diff
@@ -1,18 +1,114 @@
-from itertools import groupby
+from itertools import groupby, product
 
 import cgen as c
 import numpy as np
 from cached_property import cached_property
 
-from devito.ir.iet import (Expression, Iteration, List, FindAdjacent,
-                           FindNodes, IsPerfectIteration, Transformer,
-                           compose_nodes, retrieve_iteration_tree)
+from devito.ir.iet import (Expression, Iteration, List, FindAdjacent, FindNodes,
+                           IsPerfectIteration, Transformer, PARALLEL, AFFINE, make_efunc,
+                           compose_nodes, filter_iterations, retrieve_iteration_tree)
 from devito.logger import warning
 from devito.symbolics import as_symbol, xreplace_indices
 from devito.tools import as_tuple, flatten
 from devito.types import IncrDimension, Scalar
 
-__all__ = ['BlockDimension', 'fold_blockable_tree', 'unfold_blocked_tree']
+__all__ = ['Blocker', 'BlockDimension']
+
+
+class Blocker(object):
+
+    def __init__(self, blockinner, blockalways):
+        self.blockinner = bool(blockinner)
+        self.blockalways = bool(blockalways)
+
+        self.nblocked = 0
+
+    def make_blocking(self, iet):
```

Comment:
note for reviewers: 99% of this has just been lifted from rewriters into here -- nothing really new aside from ``self.nblocked``